### PR TITLE
Fix node pools on master

### LINKED ISSUE #261: Fix node pools on master
I believe the node pools feature is broken since merging #226.
Please consider not to use binaries built from master if you'd like to use node pools.
See https://github.com/coreos/kube-aws/pull/226#issuecomment-273343076 for more information.

### PR DESCRIPTION
**Pull request by mumoshu**
This is a follow up for the PR #226 "Don't precalculate etcd static IP addresses"

To see why this follow up was needed, read the maintainer's comment at #226 (comment)

P.S. As the result of some refactorings done in order to implement this, worker nodes powered by a spot fleet now supports the `awsEnvironment` experimental feature

fixes #261 

cc @redbaron

### CODE REVIEW

**Comment by mumoshu** [comment]
@redbaron Thanks for taking your time reviewing this!
> Output json list of IP addresses which pool stack then import and use

Wouldn't it break TLS between etcd and node pools then?

File: config/templates/cloud-config-worker [hunk 1]
```diff
@@ -29,8 +29,8 @@ coreos:
           {{.AWSCliImageRepo}}:{{.AWSCliTag}} -- cfn-init -v \
               -c "etcd-client" \
               --region {{.Region}} \
-              --resource LaunchConfigurationController \
-              --stack {{ .ClusterName }}
```
Comment (mumoshu) [inline config/templates/cloud-config-worker]: Sounds great!
However `cfn-init` doesn't seem to have access to a stack other than the one in which the EC2 instance running `cfn-init` is directly/indrectly managed 😢 
`cfn-init` fails like:

```
Jan 18 23:46:32 ip-10-0-1-242.ap-northeast-1.compute.internal rkt[7989]: [ 1044.115954] awscli[5]: ValidationError: Instance i-0507f6146f3c1bd79 in not a member of stack arn:aws:cloudformation:ap-northeast-1:<account-id>:stack/kubeawstest2/4dfa4280-dd8e-11e6-8dbb-500c44f24ce6
```

**Comment by redbaron** [comment]
~private ip addresses~ , etcd DNS names

**Comment by mumoshu** [comment]
Thanks! So your intention is to revive `etcdEndpoints` key in the nodepool's `cluster.yaml` or [what I've done in this work to allow a node pool to import etcd dns names from a main cluster](https://github.com/coreos/kube-aws/pull/263/files#diff-4c2a1b47467158c986623aa3a3d3c323R1244)?

**Comment by redbaron** [comment]
ah, nevermoind, I didn't read it close enough,you already doing export + import :) I saw changes in go code and  thought you were doing something to enable it work in nodepools, but you are  not, so it is all good.

File: config/templates/stack-template.json [hunk 2]
```diff
@@ -87,9 +87,44 @@
           {{end}}
         }
       },
-      "DependsOn" : ["AutoScaleController"]
+      "Metadata" : {
```
Comment (mumoshu) [inline config/templates/stack-template.json]: Thanks for your comment!
Regarding the mental effort, your original decision sounds very reasonable to me now.
On the other hand, I'd also like only relevant configsets to be included in the metadata of a AWS resource like an ASG for workers and an ASG for controllers, respectively.
So, would you think it is not the best but acceptable/slightly better to extract each configset to a golang subtemplate like:

```yaml
{{define "cfn-init-configset-aws-environment"}}
{
  "commands": {
  "write-environment": {
  "command": { "Fn::Join" : ["", [ "echo '",
  {{range $variable, $function := .Experimental.AwsEnvironment.Environment}}
  "{{$variable}}=", {{$function}} , "\n",
  {{end}}
  "' > /etc/aws-environment" ] ] }
  }
}
{{end}}
{{define "cfn-init-configset-etcd-client"}}
...
{{end}}
{{define "cfn-init-configset-etcd-server"}}
...
{{end}}
```

Comment (redbaron) [inline config/templates/stack-template.json]: nice and clean,I like it

**Pull request merged**